Reload part of the server configuration without restarts

SomeoneToIgnore opened the pull request:
Partially addresses https://github.com/rust-analyzer/rust-analyzer/issues/2857

Closes #3751

Reloads all server configuration that's not related to VFS without restarts.

The VFS-related parameters are not considered, since VFS is planned to be rewritten/replaced in the future and I have a suspicion that with the current code, swapping the VFS and the file watchers on the fly will cause big troubles.

I have to store and process the config request id separately, since the `workspace/configuration` response returns `any[]` (https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration), if there's a better way to handle those responses, let me know.

edwin0cheng commented on crates/rust-analyzer/src/world.rs:
```diff
@@ -159,20 +176,7 @@ impl WorldState {
         change.set_crate_graph(crate_graph);
 
         // FIXME: Figure out the multi-workspace situation
```
Have we already works in multi-workspace ?

SomeoneToIgnore replied:
Indeed, thanks, I'll remove this fixme a bit later.

Veetaha commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -389,19 +399,35 @@ fn loop_turn(
                 }
+
+                if Some(resp.id) == loop_state.configuration_request_id {
+                    loop_state.configuration_request_id.take();
```
```suggestion
                    loop_state.configuration_request_id = None;
```

SomeoneToIgnore replied:
Looks better indeed, thanks.

Veetaha commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -389,19 +399,35 @@ fn loop_turn(
+                            get_options(&new_config, text_document_caps),
+                            get_feature_flags(&new_config, connection),
+                        );
+                    }
```
Avoid unwrap?
```suggestion
                    } else if let Some(result) = resp.result {
                        let new_config =
                            serde_json::from_value::<Vec<ServerConfig>>(result)?
                                .first()
                                .expect(
                                    "The client is expected to always send a non-empty config data",
                                )
                                .to_owned();
                        world_state.update_configuration(
                            new_config.lru_capacity,
                            get_options(&new_config, text_document_caps),
                            get_feature_flags(&new_config, connection),
                        );
                    } else {
                        log::error!("received empty server settings response from the client")
                    }
```

SomeoneToIgnore replied:
I check this before unwrapping:

```rust
else if resp.result.is_none() {
    log::error!("received empty server settings response from the client")
}
```

Veetaha replied:
Sure, it is clearly seen that you check for `is_none()`, but I prefer not having the `unwrap()` and relying on the type system, so that you avoid the space for the errors like removing that `is_none()` check by-mistake and not checking the code after it. I remember you were talking about avoiding unwraps too https://github.com/rust-analyzer/rust-analyzer/pull/3623#discussion_r393600230
```rust
// Prefer:
if let Some(val) = opt {
    ...
} else {
    ...
}
// To
// 
if opt.is_none() {
    ...
} else {
    let opt = opt.unwrap();
}
```

Veetaha commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -624,7 +649,17 @@ fn on_notification(
+                ConfigurationParams::default(),
+            );
+            msg_sender.send(request.into())?;
+            loop_state.configuration_request_id.replace(request_id);
```
```suggestion
            loop_state.configuration_request_id = Some(request_id);
```

Veetaha commented on editors/code/src/client.ts:
```diff
@@ -5,6 +5,30 @@ import { Config } from './config';
 import { CallHierarchyFeature } from 'vscode-languageclient/lib/callHierarchy.proposed';
 import { SemanticTokensFeature, DocumentSemanticsTokensSignature } from 'vscode-languageclient/lib/semanticTokens.proposed';
 
+export function configToServerOptions(config: Config): object {
```
```suggestion
export function configToServerOptions(config: Config) {
```
I'd suggest not doing this upcast here, let the type inference make its job...

Veetaha commented on editors/code/src/inlay_hints.ts:
```diff
@@ -14,6 +14,8 @@ export function activateInlayHints(ctx: Ctx) {
             }
             if (!this.updater) this.updater = new HintsUpdater(ctx);
+
+            this.updater.syncCacheAndRenderHints();
```
```suggestion
            if (this.updater) {
                this.updater.syncCacheAndRenderHints();
            } else {
                this.updater = new HintsUpdater(ctx);
            }
```
The new updater will sync cache and render hints in it's constructor, so it's better to avoid doing it 2 times.

Veetaha commented on editors/code/src/inlay_hints.ts:
```diff
@@ -104,7 +106,7 @@ class HintsUpdater implements Disposable {
     }
 
-    private syncCacheAndRenderHints() {
+    public syncCacheAndRenderHints() {
```
```suggestion
    syncCacheAndRenderHints() {
```

Veetaha commented on editors/code/src/main.ts:
```diff
@@ -93,6 +93,12 @@ export async function activate(context: vscode.ExtensionContext) {
+        _ => ctx?.client?.sendNotification('workspace/didChangeConfiguration', { settings: "" }),
+        null,
+        ctx?.subscriptions,
+    );
```
```suggestion
    vscode.workspace.onDidChangeConfiguration(
        _ => ctx.client.sendNotification('workspace/didChangeConfiguration', { settings: "" }),
        null,
        ctx.subscriptions,
    );
```
TypeScipt has context typing, thus if you hover over `ctx` type here, you should not see `| undefined`.

I also wonder if we could just mass the configuration here, but then we won't follow the lsp protocol, won't we?
I wish `vscode-languageclient` could automatically send the `workspace/didChangeConfiguration` for us...

SomeoneToIgnore replied:
> TypeScipt has contextual typing, thus if you hover over ctx type here, you should not see | undefined.

Only the `ctx.subscriptions,` compiles, the closure expression requires `?`
<img width="772" alt="image" src="https://user-images.githubusercontent.com/2690773/77237968-2b7c6480-6bd5-11ea-863d-25b2bdd8b311.png">

> I also wonder if we could just pass the configuration here, but then we won't follow the lsp protocol, will we?

No, we should not, I've left the comment on the Rust counterpart: 
https://github.com/rust-analyzer/rust-analyzer/pull/3666/files#diff-688b34883e8efae680c3c682383af7e3R653

Veetaha replied:
Ah, right, this is a closure, and a closure can be called at any point of time, including when `ctx === undefined`.
But however, you can remove that `client?.` question mark at least, anyway.

kiljacken commented on crates/rust-analyzer/src/world.rs:
```diff
@@ -158,21 +175,7 @@ impl WorldState {
             });
         change.set_crate_graph(crate_graph);
 
-        // FIXME: Figure out the multi-workspace situation
```
This fixme has not actually been adressed, would you mind moving it into `create_watcher`?

SomeoneToIgnore replied:
Sure thing, moved.

SomeoneToIgnore commented:
I would really appreciate if somebody can review and merge this before the https://github.com/rust-analyzer/rust-analyzer/pull/3696 lands :)

Maybe @matklad ?

Veetaha commented:
Ag, sorry, there wil be a conflict ;)

Veetaha commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -389,19 +399,32 @@ fn loop_turn(
+                    if let Some(err) = resp.error {
+                        log::error!("failed to fetch the server settings: {:?}", err)
+                    } else if let Some(result) = resp.result {
+                        let new_config = serde_json::from_value::<Vec<ServerConfig>>(result)?
```
Does this have to be a `Vec`? I don't see that the lsp protocol says it has to be so, its just `any`, not `any[]`

SomeoneToIgnore replied:
> it's just any, not any[]

It's `any[]`: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration

```
Response:

    result: any[]
    error: code and message set in case an exception happens during the ‘workspace/configuration’ request
```

Veetaha replied:
Ah, yeah, looking at the wrong thing...

edwin0cheng commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -389,19 +399,32 @@ fn loop_turn(
+
+                if Some(resp.id) == loop_state.configuration_request_id {
+                    loop_state.configuration_request_id = None;
+                    if let Some(err) = resp.error {
```
A little nit, how about :

```rust
Response { error, result, ... } = resp; 
match (error, result) {
```

SomeoneToIgnore replied:
Sure.

edwin0cheng commented:
I just tested, at least `"rust-analyzer.highlightingOn"` do not works (No reload notification or do not change any highlight)

matklad commented:
Highlighting may be broken for some other reason. I think we should just
delete old, non semantic-tokens based code already.

On Tue, 24 Mar 2020 at 15:51, Edwin Cheng <notifications@github.com> wrote:

> I just tested, at least "rust-analyzer.highlightingOn" do not works. (No
> reload notification or do not change any highlight)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-analyzer/rust-analyzer/pull/3666#issuecomment-603284283>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANB3M2T3TQ4COZEGKUA46DRJDCFRANCNFSM4LQV3ECA>
> .
>

edwin0cheng commented:
@matklad Agreed. 

OTHO , I just tested:

* inlay hints and parameter hints that both works correctly.
* `rust-analyzer.cargo-watch.enable` works, but if I disable it during cargo checking, the progress bar display in VSCode will not clear.

matklad commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -389,19 +399,32 @@ fn loop_turn(
+                    if let Some(err) = resp.error {
+                        log::error!("failed to fetch the server settings: {:?}", err)
+                    } else if let Some(result) = resp.result {
+                        let new_config = serde_json::from_value::<Vec<ServerConfig>>(result)?
```
This `?` will crush the server if config is invalid. We should ignore invalid configs instead.

SomeoneToIgnore replied:
Thanks, fixed.

SomeoneToIgnore commented:
Thanks a lot for testing, nice catches.

I've moved the whole `highlighting` section under the reload check, so the prompt now appears for all of them.
If I got that right, with semantic tokens feature on, we highlight the documents unconditionally, so that setting is part of the old highlighting mechanism.
I wonder, should we provide something similar to the new one too?

As for the check interruption, this gets interesting: it looks like we try to do the graceful shutdown in the `Drop` impl for the `CheckWatcher`.
I have no clue why does not it finish correctly though, since don't know how it interacts with VS Code.
Might it be that we need to send some response to VS Code on termination also?

eriktrom commented:
@SomeoneToIgnore - never used vscode framework, but give this a shot, in editors/code/src/config.ts, [line 77](https://github.com/rust-analyzer/rust-analyzer/pull/3666/commits/fe379be8f38225a6c8b296c818698a37637bc16d#diff-519236f72123ff858d99b50e506bf6ccL77), `return Promise.resolve()` so that when u pass the callback to the constructor function's first ev listener(in that file), `this.refreshConfig()` is always scheduled in the microtask queue where promises resolve. U should then also be able to remove `this.refreshConfig();` from the 2nd line in the constructor, thus forcing it to be called in the same event loop stack frame.

I've noticed this bug, so if you need help, holler, don't mind helping out.

(and of course, i could be way wrong and the scheduling could very well be benign)

Veetaha commented:
@eriktrom what bug are you talking about? What do you mean by passing the callback to the event listener? Maybe you meant registering the event listener callback which is this line:
https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/editors/code/src/config.ts#L32
And I don't see any bugs and how microtasks should be connected with them here.

@SomeoneToIgnore if you feel that there is something that is broken on the frontend that doesn't support the reloading, you can pass this task to me. Anyway, my guess is that the server doesn't send the `end` event when it interrupts the cargo watcher. So this line here ends with an error 
https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L381

because the communication channel was closed abruptly in the `WatchThread::drop()`
https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L398

From my perspective, it looks like the proper solution is to add a `Shutdown` command here:
https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L98-L101

And properly handle it here by sending the `WorkDoneProgressEnd` task:
https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L123-L129

Or maybe we should not drop the cargo watcher, so that we don't send the `Shutdown` command in drop, but maybe instead just stop the current check process by sending the new `Update` command. However, we need to add an API to update the watcher config.

eriktrom commented:
sorry let me clarify succinctly:

https://github.com/rust-analyzer/rust-analyzer/pull/3666/commits/fe379be8f38225a6c8b296c818698a37637bc16d#diff-519236f72123ff858d99b50e506bf6ccL56-L57

https://github.com/rust-analyzer/rust-analyzer/pull/3666/commits/fe379be8f38225a6c8b296c818698a37637bc16d#diff-519236f72123ff858d99b50e506bf6ccL70-L87.patch

in the constructor. the callback  `this.onConfigChange`, registered on line 56 by the `vscode.workspace.onDidChangeConfiguration(this.onConfigChange, this, ctx.subscriptions);` event handler is *async*. But when called, due to being an async callback, we're already in a different turn of the event loop. So line 57 is being called immediately, and the method `this.onConfigChange` is non-deterministically called by the event handler (registered on line 56), I'd need to debug with a breakpoint but a few things can go wrong, the instance could be destroyed in that time period, we might then re-register that async callback if its a hot path, preventing the normal sync stack from flushing, (b/c promises block while they're queue has any microtasks), leading to a timeout, that then requires an host extension restart.

I'm not sure how vscode does its internal scheduling, so I just commented offhand as though it didn't and therefore, take with grain of salt.

BTW - I love rust-analyzer :) thanks so much for ur work.

eriktrom commented on editors/code/src/config.ts:
```diff
@@ -44,7 +53,7 @@ export class Config {
     constructor(private readonly ctx: vscode.ExtensionContext) {
-        vscode.workspace.onDidChangeConfiguration(this.refreshConfig, this, ctx.subscriptions);
+        vscode.workspace.onDidChangeConfiguration(this.onConfigChange, this, ctx.subscriptions);
         this.refreshConfig();
```
here is my concern - refreshConfig is sync, onConfigChange is async, fyi

SomeoneToIgnore replied:
Thank you for rising it.
It looks like we use async event handlers here and in the old buggy highlighting module, so this sounds suspicious indeed.

Yet I agree with [this idea](https://github.com/rust-analyzer/rust-analyzer/pull/3666#issuecomment-603600264) for the particular case and willing to try it first.

As for `async` in such cases, @Veetaha definitely knows more about about this so might be worth to discuss it with him.

eriktrom replied:
np, just noticed - too much nodejs in my blood to not :)

Veetaha replied:
@eriktrom

> in the constructor. the callback `this.onConfigChange`, registered on line 56 by the `vscode.workspace.onDidChangeConfiguration(this.onConfigChange, this, ctx.subscriptions);` event handler is async. But when called, due to being an async callback, we're already in a different turn of the event loop.

This is not entirely the truth. Promises are executed eagerly, but resolved lazily (i.e. at minimum when the current call stack is emptied). The following code proves it
<details/>
<summary>Code itself</summary>

```ts
async function foo() { ++i; }

let i = 0;

foo().then(() => console.log("resolved: " + i));

console.log("After foo(): " + i);
```
</details>

![image](https://user-images.githubusercontent.com/36276403/77562371-99bc7200-6ec8-11ea-8fb4-1cd3688dfbf3.png)

As you see, the variable gets incremented, right after `foo()` because it is executed eagerly (i.e. synchronously).

> So line 57 is being called immediately, and the method this.onConfigChange is non-deterministically called by the event handler (registered on line 56),  I'd need to debug with a breakpoint but a few things can go wrong, the instance could be destroyed in that time period
we might then re-register that async callback if its a hot path, preventing the normal sync stack from flushing, (b/c promises block while they're queue has any microtasks), leading to a timeout, that then requires an host extension restart.

Right, but "*destroyed*" is not the proper term since the lifetime of objects in JavaScript is indeterminate. Properly speaking, the instance can be logically dropped off, but the event listener is not deregistered.

So, the right way to solve it is to make config `implements Disposable` and deregister the event listener when the config is disposed of. *Right now* this is not a big problem.

It may happen only when the user invokes `Restart server` command. Because when doing it we create a new config, but the previous config's event listener is not deregistered.

Anyway, this is not connected with this particular PR at all, but thanks for spotting this!

eriktrom replied:
yeah let's not derail - last note, no response needed, sorry 'bout bringing this up!

```
> async function foo() { i++; return; }
undefined
> foo()
Promise { undefined }
> async function foo() { i++; return Promise.resolve(); }
undefined
> foo()
Promise { <pending> } **<-- WARNING, i can haz bug**
```
the first is immediately resolved, the 2nd waits for an event. the callback in question has an if statement that can return early, returning an unwrapped promise. tl;dr

such fun right? over and out

kiljacken commented:
> Anyway, my guess is that the server doesn't send the `end` event when it interrupts the cargo watcher. So this line here ends with an error
> 
> https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L381
> 
> because the communication channel was closed abruptly in the `WatchThread::drop()`
> 
> https://github.com/rust-analyzer/rust-analyzer/blob/6ad1a0711631d8017791a6dfe85bbe205d6c7414/crates/ra_cargo_watch/src/lib.rs#L398

I believe your analysis is correct. I didn't really consider cancelling the check thread without starting a new run immediately afterwards which masks the underlying issue. I believe we are robust against multiple `end` messages in the extension, so just making sure we send it in a shutdown command or something else should be fine.

edwin0cheng commented:
> And properly handle it here by sending the WorkDoneProgressEnd task:

@Veetaha @SomeoneToIgnore One thing I want to clarify here is that the `Cargo` progress bar is a custom progress bar, not the default VSCode progress bar.

matklad commented on editors/code/src/main.ts:
```diff
@@ -93,6 +93,12 @@ export async function activate(context: vscode.ExtensionContext) {
     activateInlayHints(ctx);
+
+    vscode.workspace.onDidChangeConfiguration(
+        _ => ctx?.client?.sendNotification('workspace/didChangeConfiguration', { settings: "" }),
```
Hm, is there perhaps some way to make the client library we use do this automatically?

matklad commented on editors/code/src/ctx.ts:
```diff
@@ -20,6 +20,7 @@ export class Ctx {
         const res = new Ctx(config, extCtx, client, serverPath);
         res.pushCleanup(client.start());
         await client.onReady();
+        client.onRequest('workspace/configuration', _ => [configToServerOptions(config)]);
```
Similarly, is there some kind of build-in handler here, which just fetches the relevant section from the config?

matklad commented on crates/rust-analyzer/src/main_loop.rs:
```diff
@@ -624,7 +661,17 @@ fn on_notification(
         Err(not) => not,
     };
     let not = match notification_cast::<req::DidChangeConfiguration>(not) {
-        Ok(_params) => {
```
Hm, yeah, given that we already have no-op handler to avoid errors, I think clients *do* send this notification already?

SomeoneToIgnore replied:
It did not trigger for me when I tried to change the settings on `master`, so it looks like they _don't_.

But I find this surprising also, I would have expected it to be working by default.

SomeoneToIgnore commented:
I've finally found some time to dig and fix the progress bar issue.

We cannot use the way discussed at https://github.com/rust-analyzer/rust-analyzer/pull/3666#issuecomment-603729038 due to the way the main loop [works](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/rust-analyzer/src/main_loop.rs#L235): it uses the `task_recv` from the current check watcher.
This means that we cannot send whatever message from the old watcher and just drop it: the message won't be received by anybody, hence no progress bar cleaning will be done ever.

Instead, I clean up all diagnostics and stop the progress bad on `CheckWatcherThread::run` which is done once when the new watcher starts.



With this fixed, I've addressed all feedback up to [here](https://github.com/rust-analyzer/rust-analyzer/pull/3666#issuecomment-603767445).

The last portion of the feedback remaining is the one from @matklad about the automatic config requests/responses handling, the one that I've outlined in the PR header as

> if there's a better way to handle those responses, let me know.

I still have no idea if there's a better way and will gladly put a `fixme` and fix it later instead of dealing with it now and holding the PR for whatever uncertain amount of time more.

matklad commented:
bors r+
